cloudinit/distros/__init__.py

```diff
@@ -83,6 +83,30 @@ def __init__(self, name, cfg, paths):
         self.name = name
         self.networking = self.networking_cls()
 
+    def __setstate__(self, state):
+        """Restore instance state and handle missing attributes on upgrade.
+
+        This will be called when an instance of this class is unpickled; the
+        previous instance's `__dict__` is passed as `state`.  This method
+        restores that state into the current instance, and fixes up any
+        missing/incorrect state (usually caused by upgrades to cloud-init since
+        `state` was pickled).
+
+        See https://docs.python.org/3/library/pickle.html#object.__setstate__
+        for further background.
+        """
+        # Restore instance attributes
+        self.__dict__.update(state)
+        if "networking" not in self.__dict__ or not self.networking.__dict__:
```

Comment:
in retrospect, or seems painfully obvious to put it here on the Distro class, rather than the networking class

Reply:
Is there a reason for choosing __dict__ versus __class__ or some other internal attribute of an instantiated object?

Reply:
Yes: `__dict__` is "A dictionary or other mapping object used to store an object’s (writable) attributes." per https://docs.python.org/3/library/stdtypes.html#object.__dict__, which is to say that it's where instance state is stored. We only want to replace `self.networking` if it has _no_ instance state: if it already has instance state then we would be discarding that existing state, which we don't want to do.

Also: if `self.networking` already has instance state at this point (immediately after unpickle) then that means that its pickled representation must have included the `BUILD` opcode: absent a custom `__new__` implementation, that's the only way that state gets set. If the `BUILD` opcode is present, then `Networking.__setstate__` will be called if implemented. So: if there is already state (which there will be for every pickle created after that `blacklist_drivers` change landed), then we can handle any upgrade-related state remediation in `Networking` itself rather than needing to do it here.